Improve yaml parse error handling

Br4nd0R opened the pull request:
I was unable to find an attribute on `ScannerError` for the name of the file being parsed.  I included "likely cumulusci.yml" in the error messaging, because we can't be sure which call to `handle_exception()` is being invoked.

# Changes
* CumulusCI displays more user friendly error message when encountering parsing errors in `cumulusci.yml`.

davisagli commented on cumulusci/cli/cci.py:
```diff
@@ -272,6 +279,17 @@ def connection_error_message():
 
 
+def yaml_parse_error_message(error):
+    message = f"An error occurred while parsing a yaml file (likely cumulusci.yml). Check around line {error.problem_mark.line} for the issue."
```
I think the cci_safe_load function in cumulusci.utils.yaml.cumulusci_yml would be a better place to catch parse errors and reraise them with a better message -- then we should be able to get the filename. We should also include the column if it's available, and include the original error message for those who can interpret it.

davisagli commented on cumulusci/utils/yaml/tests/test_cumulusci_yml.py:
```diff
@@ -37,3 +42,39 @@ def test_converter_is_selective(self):
+        assert (
+            "An error occurred parsing cumulusci.yml at line 12345, column 54321.\nError message: None"
+            == error.value.args[0]
+        )
```
Could you rewrite this test to use an actual poorly formatted yaml file in a tempdir instead of mocks? The mocks make it hard to know if we're setting up the mocks realistically.

Also do a test to make sure cci_safe_load still works if we pass in a StringIO() instance instead of a file, because I think they probably don't have a `.name`

davisagli commented on cumulusci/utils/yaml/cumulusci_yml.py:
```diff
@@ -35,7 +39,18 @@ def _replacer_func(matchobj):
-    rc = yaml.safe_load(StringIO(data))
+    try:
+        rc = yaml.safe_load(StringIO(data))
+    except ScannerError as e:
```
Looking at the PyYAML source, there are a few classes of exceptions which extend MarkedYAMLError. I think we probably want to catch that one here instead of only ScannerError.

prescod reviewed (commented):
I seem to usually fall on the side of "mocks if necessary but only if necessary" so I like the recent changes to the PR. Libraries can change under us, so it is nice to test that they still behave the way we expect (in this case, that PyYaml still responds as we expect to typos).

I try to use shorter snippets of error messages so that minor rewordings don't break unit tests but that's hardly urgent.

Worth considering if there are other weird cases worth testing. In particular 1. What if it isn't really Unicode and 2. What if the file is totally empty.

davisagli commented on cumulusci/utils/yaml/tests/test_cumulusci_yml.py:
```diff
@@ -37,3 +43,59 @@ def test_converter_is_selective(self):
+                with pytest.raises(CumulusCIException) as error:
+                    cci_safe_load(cumulusci_yml)
+
+        assert error.typename == "CumulusCIException"
```
This is unnecessary -- the `pytest.raises` context manager already asserts that the expected exception was raised. You can also pass `matches` to `pytest.raises` to assert that a string is included in the except message, instead of a separate assertion, if you want.

davisagli reviewed (approved):
Would be good to also add tests for the other cases Paul came up with, but this is already a good improvement.

Br4nd0R commented:
@davisagli @prescod - I removed the class to make things more pytest-like (also added a fixture to clean up cumulusci.yml file creation). Odd though, ass the `test_converter()` method now fails. The `outp` has 12 spaces between `\n` and `y:` where as the value for `rc` has 10 spaces between `\n` and `y:`, but no code changed there.

davisagli commented:
@Br4nd0R fwiw, I don't see organizing tests into classes as antithetical to pytest's approach -- but I don't feel strongly about whether we do so or not.

The failing test is because you changed the indentation inside the triple-quoted string.

Br4nd0R commented:
> The failing test is because you changed the indentation inside the triple-quoted string.


![](https://media0.giphy.com/media/3o7btUg31OCi0NXdkY/giphy.gif)
